text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,542 @@
+in-memory representation, set within the returned context object. 
+
+**`HTTP_Injector(context, headers) -> headers`**  
+Each concern must implement an HTTP_Injector, which can take the in-memory 
+representation of its data from the given context object, and add it to an 
+existing set of HTTP headers.
+
+### Optional: Global Propagators
+It is often convenient to create a chain of propagators during program 
+initialization, and then access these combined propagators later in the program. 
+To facilitate this, global injectors and extractors are optionally available. 
+However, there is no requirement to use this feature.
+
+**`GetExtractors() -> []http_extractor`**  
+To access the global extractor, the Propagation API provides a function which 
+returns an extractor.
+
+**`SetExtractors([]http_extractor)`**  
+To update the global extractor, the Propagation API provides a function which 
+takes an extractor.
+
+**`GetInjectors() -> []http_injector`**  
+To access the global injector, the Propagation API provides a function which 
+returns an injector.
+
+**`SetInjectors([]http_injector)`**  
+To update the global injector, the Propagation API provides a function which 
+takes an injector.
+
+# Prototypes
+
+**Erlang:** https://github.com/open-telemetry/opentelemetry-erlang-api/pull/4  
+**Go:** https://github.com/open-telemetry/opentelemetry-go/pull/297  
+**Java:** https://github.com/open-telemetry/opentelemetry-java/pull/655  
+**Python:** https://github.com/open-telemetry/opentelemetry-python/pull/278  
+**Ruby:** https://github.com/open-telemetry/opentelemetry-ruby/pull/147  
+
+# Examples
+
+It might be helpful to look at an example, written in pseudocode. Let's describe 
+a simple scenario, where `service A` responds to an HTTP request from a `client` 
+with the result of a request to `service B`.
+
+```
+client -> service A -> service B
+```
+
+Now, let's assume the `client` in the above system is version 1.0. With version 
+v2.0 of the `client`, `service A` must call `service C` instead of `service B` 
+in order to return the correct data.
+
+```
+client -> service A -> service C
+```
+
+In this example, we would like `service A` to decide on which backend service 
+to call, based on the client version. We would also like to trace the entire 
+system, in order to understand if requests to `service C` are slower or faster 
+than `service B`. What might `service A` look like?
+
+## Global initialization 
+First, during program initialization, `service A` might set a global 
+extractor and injector which chains together baggage and tracing 
+propagation. Let's assume this tracing system is configured to use B3, 
+and has a specific propagator for that format. Initializating the propagators 
+might look like this:
+
+```php
+func InitializeOpentelemetry() {
+  // create the propagators for tracing and baggage.
+  bagExtract, bagInject = Baggage::HTTPPropagator()
+  traceExtract, traceInject = Tracer::B3Propagator()
```

Comment:
Would you say in this case that the **Tracer instance** provides an alternate implementation of B3 in this code. Should "B3Propagator" be a method on the SDK type? Should "B3Propagator" be a method on the API interface?

Reply:
The SDK always needs to provide the implementation, as the internal details go beyond which headers to use (B3 vs Trace-Context) and include filling in the values from the SDKs in-memory representation. So these propagators cannot be an API-only concept. 

As for adding constructors to the API for "standard" propagator types, so that they can be constructed at a call site without knowing which SDK is in use? I'd want to see that we needed it first. 

It feels backwards compatible to add these constructors later, and say that propagator choice is an SDK concern. That may be enough; it's simple to store propagators in a value, and flow them into places where they need to be.

```
b3Propagator = SDK::B3Propagator()
traceContextPropagator = SDK::TraceContextPropagator()

initClientA(b3Propagator)
initClientB(traceContextPropagator)
```

Reply:
But _which_ SDK  fills in values from the in-memory representation?  Since we support multiple SDKs in the process it's not clear.  The same point came up here: https://github.com/open-telemetry/oteps/issues/73

Spans have a current Tracer they can use to fill in those details, but Span Contexts do not. So, I'm not sure about this.

Reply:
Perhaps this just a pseudocode problem. I take it `SDK::TraceContextPropagator` refers to an _instance_ of the SDK, i.e., this is the moment where we bind a Tracer to a propagator and install it.

Reply:
If you agree, feel free to resolve this.